docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,329 @@
+
+We can use the GitHub API and the user permissions to the repository as the authorization mechanism. Beware the GitHub API has a [rate limit](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/rate-limits-for-github-apps#installation-access-tokens-on-githubcom), so we may want to cache the permissions for a while if we foresee hitting these limits. Also, see the [List collaborators endpoint](https://docs.github.com/en/free-pro-team@latest/rest/collaborators/collaborators?apiVersion=2022-11-28#list-repository-collaborators) for more information on authorization.
+
+### Sign Up and Sign In
+
+We can use the GitHub app as an OAuth provider for Wing Cloud. A set of pages will be needed:
+
+```mermaid
+sequenceDiagram
+    wing.cloud/login->>Wing Cloud Platform GitHub App: Redirect
+    Wing Cloud Platform GitHub App-->>Wing Cloud Platform GitHub App: Login to GitHub
+    Wing Cloud Platform GitHub App-->>wing.cloud/login/callback: Redirect (with some tokens)
+    wing.cloud/login/callback-->>wing.cloud/login/callback: Generate a new user ID on our tables, if it doesn't exist
+    wing.cloud/login/callback-->>wing.cloud/login/callback: Exchange tokens and encode them in a cookie for https://wing.cloud
+    wing.cloud/login/callback-)wing.cloud: Redirect (now with the authenticated cookie)
+```
+
+> **Important Note:** Since we don't ask for information during the sign-up process, both the sign-up and sign in flows are the same.
+
+### Wing Cloud Dashboard
+
+1. get user repositories from GitHub by using the GitHub API and the user's access token we have in our DB
+2. filter repositories that have Wing Cloud application installed.
+3. for each repository, get all preview environments.
+4. there might be need to store repository -> preview environments mapping in our DB.
+5. there is no need to store user -> repositories mapping in our DB, we can get it from GitHub API.
+
+### Processing GitHub Events
+
+A Cloud.Function will listen to GitHub events and will process them. Based on the event, the function will create, update or delete preview environments on Flyio.
+
+```mermaid
+graph LR
+    gh{GitHub}
+    ghapp>GitHub App]
+    fn[Process Events Fn]
+    flyio{Flyio}
+    gh -- Integration --> ghapp
+    ghapp -- Events --> fn
+    fn -- Create Preview Environment --> flyio
+    fn -- Update Preview Environment --> flyio
+    fn -- Delete Preview Environment --> flyio
+```
+
+Depending on whether GitHub retries events or not, we may want to include a FIFO queue in the middle.
+
+```mermaid
+graph LR
+    gh{GitHub}
+    ghapp>GitHub App]
+    fn[Process Events Fn]
+    queue[(Events Queue)]
+    fn2[Enqueue Events Fn]
+    ghapp -- Events --> fn2
+    fn2 --> queue
+    queue --> fn
+    gh -- Integration --> ghapp
+    fn --> rest[...]
+```
+
+The probot npm package will take care of ensuring the events received are legitimate.
```

Comment:
What part of the `github -> wingcloud` flow will be managed by the probot and which part by the wing project?